secret-manager/runtime/src/main/java/io/quarkiverse/googlecloudservices/secretmanager/runtime/config/SecretManagerConfigSource.java

```diff
@@ -0,0 +1,48 @@
+package io.quarkiverse.googlecloudservices.secretmanager.runtime.config;
+
+import java.util.Collections;
+import java.util.Map;
+
+import com.google.cloud.secretmanager.v1.AccessSecretVersionResponse;
+import com.google.cloud.secretmanager.v1.SecretManagerServiceClient;
+import com.google.cloud.secretmanager.v1.SecretVersionName;
+
+import io.quarkus.arc.Arc;
+import io.smallrye.config.common.AbstractConfigSource;
+
+public class SecretManagerConfigSource extends AbstractConfigSource {
+
+    private static final String CONFIG_SOURCE_NAME = "io.quarkiverse.googlecloudservices.secretmanager.runtime.config";
+
+    private final String defaultProject;
+
+    // This will be initialized the first time `getValue` is called.
+    private SecretManagerServiceClient client = null;
+
+    public SecretManagerConfigSource(String defaultProject) {
+        super(CONFIG_SOURCE_NAME, DEFAULT_ORDINAL);
+        this.defaultProject = defaultProject;
+    }
+
+    @Override
+    public Map<String, String> getProperties() {
+        return Collections.emptyMap();
+    }
+
+    @Override
+    public String getValue(String propertyName) {
+        SecretVersionName secretVersionName = SecretManagerConfigUtils.getSecretVersionName(propertyName, defaultProject);
+        if (secretVersionName == null) {
+            // The propertyName is not in the form "${sm//...}" so return null.
+            return null;
+        }
+
+        if (client == null) {
+            // Retrieve the Secret Manager client in the context.
+            client = Arc.container().instance(SecretManagerServiceClient.class).get();
+        }
```

Comment:
I'm concerned with potential race condition here, if multiple config are loaded concurently.
We better load the `SecretManagerServiceClient` inside the constructor or use some kind of concurency construct (holder class or double locked idiom)

Reply:
Done -- Introduced the wrapper class with `synchronized` guard. Unfortunately we must retrieve the instance of the client from `getValue()` as it is not available in the Quarkus application context at construction time. This is similar to what is being done in the [Vault extension for Quarkus](https://github.com/quarkusio/quarkus/blob/master/extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/config/VaultConfigSource.java).

Reply:
Vaul read it from Arc anytime: https://github.com/quarkusio/quarkus/blob/d3a7084c0b90157fa461dde5f178b3da4df27717/extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/config/VaultConfigSource.java#L88

And the wrapper class seems to be missing a volatile modifier for the client attribute for the synchronization to be correct. Synchronization is always tricky ;)

Reply:
Done! Sorry, I haven't written much multi-thread code haha.

Reply:
In fact I was wrong, `volatile` was not needed here as you synchonize at the method level so no risk that any thread make a not-yet-initialized read.
But the best for this is to use a lazy holder factory class that allows to avoid the synchronization overhead, you can have a look at this nice article explaining how it works: https://shipilev.net/blog/2014/safe-public-construction

But don't bother with it, I'll implement it later one, the current implementation is not optimal but is correct.